File: core/src/domain/sale/sale_line.py
from dataclasses import dataclass

from src.domain.book import BookDescriptor, Book
from src.domain.identity import Identity, Collection
from .sale import Sale


@dataclass
class SaleLine(Identity):
    """
    SaleLine entity represents a line item in a sale
    """
    sale: str = None
    book: BookDescriptor = None
    quantity: int = None
    price: float = None
    discount: float = None
    total: float = None

    @classmethod
    def create(
            cls,
            sale: Sale,
            book: Book,
            quantity: int,
            price: float,
            discount: float,
            total: float,
    ) -> "SaleLine":
        """
        Factory method to create a new SaleLine instance.
        """
        return cls(
            sale=sale.id,
            book=book.get_descriptor(),
            quantity=quantity,
            price=price,
            discount=discount,
            total=total,
        )

    def modify(
            self,
            book: Book,
            quantity: int,
            price: float,
            discount: float,
            total: float,
    ) -> None:
        """
        Method to modify the sale line details.
        """
        self.book = book.get_descriptor()
        self.quantity = quantity
        self.price = price
        self.discount = discount
        self.total = total


type SaleLineCollection = Collection[SaleLine]

File: editor/src/app/sales/model/sale-line.ts
import { BookDescriptor } from '@/books/model/book-descriptor';

export class SaleLine {
  constructor(
    public lineId: string,
    public book: BookDescriptor,
    public quantity: number,
    public price: number,
    public discount: number,
    public total: number,
  ) {}

  public static empty(): SaleLine {
    return new SaleLine('', BookDescriptor.empty(), 1, 0, 0, 0);
  }

  public static from(item: SaleLine): SaleLine {
    return new SaleLine(
      item.lineId,
      BookDescriptor.from(item.book),
      item.quantity,
      item.price,
      item.discount,
      item.total,
    );
  }
}
